File: mobile/ios/IkimonScan/Sources/API/MobilePlatformDiscoveryClient.swift
import Foundation

struct MobilePlatformDiscoveryDocument: Decodable, Equatable {
    let platform: String
    let environment: String
    let product: String
    let capabilityEndpoint: String
    let authorizationIssuer: String
    let supportedPlatformContracts: [String]
    let serverTime: String
    let descriptorDigest: String

    enum CodingKeys: String, CodingKey {
        case platform
        case environment
        case product
        case capabilityEndpoint = "capability_endpoint"
        case authorizationIssuer = "authorization_issuer"
        case supportedPlatformContracts = "supported_platform_contracts"
        case serverTime = "server_time"
        case descriptorDigest = "descriptor_digest"
    }
}

struct MobilePlatformCapability: Decodable, Equatable {
    let capabilityId: String
    let version: String
    let state: String
    let validUntil: String

    enum CodingKeys: String, CodingKey {
        case capabilityId = "capability_id"
        case version
        case state
        case validUntil = "valid_until"
    }
}

struct MobilePlatformContractRange: Decodable, Equatable {
    let min: String
    let max: String
}

struct MobilePlatformContractRanges: Decodable, Equatable {
    let platform: MobilePlatformContractRange
}

struct MobilePlatformCapabilityResponse: Decodable, Equatable {
    let maintenanceMode: String
    let contracts: MobilePlatformContractRanges
    let capabilities: [MobilePlatformCapability]
    let configDigest: String
    let validUntil: String

    enum CodingKeys: String, CodingKey {
        case maintenanceMode = "maintenance_mode"
        case contracts
        case capabilities
        case configDigest = "config_digest"
        case validUntil = "valid_until"
    }
}

struct MobilePlatformDiscoveryResult: Equatable {
    let discovery: MobilePlatformDiscoveryDocument
    let capabilities: MobilePlatformCapabilityResponse
}

enum MobilePlatformDiscoveryError: Error, Equatable {
    case invalidCapabilitiesPath
    case transportStatus(Int)
    case platformMismatch
    case environmentInvalid
    case productMismatch
    case contractMismatch
    case authorizationIssuerRequired
    case descriptorDigestInvalid
    case configDigestInvalid
    case invalidCapability
    case providerResourceLeak
}

/// Read-only client for the canonical mobile product contract.
///
/// The descriptor identifies IKIMON Cloudflare OS, while capability identifiers contain
/// no R2/D1/Queue/DO binding detail. The client therefore remains valid if backend adapters
/// are replaced and can be called from a future shared shell or a full-native client.
struct MobilePlatformDiscoveryClient {
    static let expectedPlatform = "ikimon-cloudflare-os"
    static let expectedPlatformContract = "1.0"
    static let expectedProduct = "zukan"

    let origin: URL
    let session: URLSession

    init(origin: URL = URL(string: "https://ikimon.life")!, session: URLSession = .shared) {
        self.origin = origin
        self.session = session
    }

    func discover() async throws -> MobilePlatformDiscoveryResult {
        let discoveryURL = origin.appendingPathComponent(".well-known/ikimon-platform")
        let discovery: MobilePlatformDiscoveryDocument = try await get(discoveryURL)
        try Self.validate(discovery: discovery)

        guard let capabilitiesURL = Self.resolveRelativePath(discovery.capabilityEndpoint, against: origin) else {
            throw MobilePlatformDiscoveryError.invalidCapabilitiesPath
        }
        let capabilities: MobilePlatformCapabilityResponse = try await get(capabilitiesURL)
        try Self.validate(capabilities: capabilities)
        return MobilePlatformDiscoveryResult(discovery: discovery, capabilities: capabilities)
    }

    static func validate(discovery: MobilePlatformDiscoveryDocument) throws {
        guard discovery.platform == expectedPlatform else {
            throw MobilePlatformDiscoveryError.platformMismatch
        }
        guard ["development", "staging", "production"].contains(discovery.environment) else {
            throw MobilePlatformDiscoveryError.environmentInvalid
        }
        guard discovery.product == expectedProduct else {
            throw MobilePlatformDiscoveryError.productMismatch
        }
        guard discovery.supportedPlatformContracts.contains(expectedPlatformContract) else {
            throw MobilePlatformDiscoveryError.contractMismatch
        }
        guard !discovery.authorizationIssuer.isEmpty else {
            throw MobilePlatformDiscoveryError.authorizationIssuerRequired
        }
        guard discovery.descriptorDigest.hasPrefix("sha256:") else {
            throw MobilePlatformDiscoveryError.descriptorDigestInvalid
        }
        guard isSafeRelativeAPIPath(discovery.capabilityEndpoint) else {
            throw MobilePlatformDiscoveryError.invalidCapabilitiesPath
        }
    }

    static func validate(capabilities response: MobilePlatformCapabilityResponse) throws {
        guard ["none", "read_only", "unavailable"].contains(response.maintenanceMode) else {
            throw MobilePlatformDiscoveryError.invalidCapability
        }
        guard !response.contracts.platform.min.isEmpty,
              !response.contracts.platform.max.isEmpty else {
            throw MobilePlatformDiscoveryError.contractMismatch
        }
        guard response.configDigest.hasPrefix("sha256:") else {
            throw MobilePlatformDiscoveryError.configDigestInvalid
        }
        guard !response.validUntil.isEmpty else {
            throw MobilePlatformDiscoveryError.invalidCapability
        }
        for capability in response.capabilities {
            guard !capability.capabilityId.isEmpty,
                  !capability.version.isEmpty,
                  !capability.validUntil.isEmpty,
                  ["available", "degraded", "read_only", "disabled"].contains(capability.state) else {
                throw MobilePlatformDiscoveryError.invalidCapability
            }
            if containsProviderResourceName(capability.capabilityId) {
                throw MobilePlatformDiscoveryError.providerResourceLeak
            }
        }
    }

    static func isSafeRelativeAPIPath(_ path: String) -> Bool {
        path.hasPrefix("/")
            && !path.hasPrefix("//")
            && !path.contains("://")
            && !path.contains("\\")
    }

    static func resolveRelativePath(_ path: String, against origin: URL) -> URL? {
        guard isSafeRelativeAPIPath(path),
              var components = URLComponents(url: origin, resolvingAgainstBaseURL: false),
              components.scheme == "https" || components.host == "127.0.0.1" || components.host == "localhost" else {
            return nil
        }
        components.path = path
        components.query = nil
        components.fragment = nil
        return components.url
    }

    private static func containsProviderResourceName(_ value: String) -> Bool {
        let normalized = value.lowercased()
        return ["r2_bucket", "d1_database", "durable_object", "hyperdrive_config", "queue_binding"]
            .contains { normalized.contains($0) }
    }

    private func get<T: Decodable>(_ url: URL) async throws -> T {
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("ikimon-scan/platform-discovery", forHTTPHeaderField: "User-Agent")
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw MobilePlatformDiscoveryError.transportStatus(0)
        }
        guard (200..<300).contains(http.statusCode) else {
            throw MobilePlatformDiscoveryError.transportStatus(http.statusCode)
        }
        return try JSONDecoder().decode(T.self, from: data)
    }
}

File: mobile/android/ikimon-pocket/app/src/main/kotlin/life/ikimon/api/MobilePlatformDiscoveryClient.kt
package life.ikimon.api

import android.content.Context
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.util.concurrent.TimeUnit

data class MobilePlatformDiscovery(
    val platform: String,
    val environment: String,
    val product: String,
    val capabilityEndpoint: String,
    val authorizationIssuer: String,
    val supportedPlatformContracts: List<String>,
    val serverTime: String,
    val descriptorDigest: String,
)

data class MobilePlatformCapability(
    val capabilityId: String,
    val version: String,
    val state: String,
    val validUntil: String,
)

data class MobilePlatformCapabilityResponse(
    val maintenanceMode: String,
    val platformContractMin: String,
    val platformContractMax: String,
    val capabilities: List<MobilePlatformCapability>,
    val configDigest: String,
    val validUntil: String,
)

data class MobilePlatformDiscoveryResult(
    val discovery: MobilePlatformDiscovery,
    val capabilities: MobilePlatformCapabilityResponse,
)

/**
 * Read-only discovery client for the canonical provider-resource-opaque mobile contract.
 *
 * The descriptor identifies IKIMON Cloudflare OS as the platform, while capability IDs
 * deliberately contain no R2/D1/Queue/DO binding detail. The same semantic client can be
 * reused from a future shared shell or a full-native ZUKAN client.
 *
 * Network I/O is synchronous to match the existing mobile API clients. Call from an
 * existing worker/background dispatcher rather than the Android main thread.
 */
object MobilePlatformDiscoveryClient {
    const val EXPECTED_PLATFORM = "ikimon-cloudflare-os"
    const val EXPECTED_PLATFORM_CONTRACT = "1.0"
    const val EXPECTED_PRODUCT = "zukan"

    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    fun discover(context: Context): MobilePlatformDiscoveryResult {
        val origin = MobileApiConfig.currentRuntimeOrigin(context).trimEnd('/')
        val discoveryJson = getJson("$origin/.well-known/ikimon-platform")
        val discovery = parseDiscovery(discoveryJson)
        validateDiscovery(discovery)

        val capabilityJson = getJson("$origin${discovery.capabilityEndpoint}")
        val capabilities = parseCapabilities(capabilityJson)
        validateCapabilities(capabilities)
        return MobilePlatformDiscoveryResult(discovery, capabilities)
    }

    internal fun parseDiscovery(json: JSONObject): MobilePlatformDiscovery {
        val rawContracts = json.optJSONArray("supported_platform_contracts")
        val supportedContracts = buildList {
            if (rawContracts != null) {
                for (index in 0 until rawContracts.length()) {
                    rawContracts.optString(index).takeIf(String::isNotBlank)?.let(::add)
                }
            }
        }
        return MobilePlatformDiscovery(
            platform = json.optString("platform"),
            environment = json.optString("environment"),
            product = json.optString("product"),
            capabilityEndpoint = json.optString("capability_endpoint"),
            authorizationIssuer = json.optString("authorization_issuer"),
            supportedPlatformContracts = supportedContracts,
            serverTime = json.optString("server_time"),
            descriptorDigest = json.optString("descriptor_digest"),
        )
    }

    internal fun parseCapabilities(json: JSONObject): MobilePlatformCapabilityResponse {
        val rawCapabilities = json.optJSONArray("capabilities")
        val capabilities = buildList {
            if (rawCapabilities != null) {
                for (index in 0 until rawCapabilities.length()) {
                    val item = rawCapabilities.optJSONObject(index) ?: continue
                    add(
                        MobilePlatformCapability(
                            capabilityId = item.optString("capability_id"),
                            version = item.optString("version"),
                            state = item.optString("state"),
                            validUntil = item.optString("valid_until"),
                        ),
                    )
                }
            }
        }
        val platformContracts = json.optJSONObject("contracts")?.optJSONObject("platform")
        return MobilePlatformCapabilityResponse(
            maintenanceMode = json.optString("maintenance_mode"),
            platformContractMin = platformContracts?.optString("min").orEmpty(),
            platformContractMax = platformContracts?.optString("max").orEmpty(),
            capabilities = capabilities,
            configDigest = json.optString("config_digest"),
            validUntil = json.optString("valid_until"),
        )
    }

    internal fun validateDiscovery(discovery: MobilePlatformDiscovery) {
        require(discovery.platform == EXPECTED_PLATFORM) { "mobile_platform_identity_mismatch" }
        require(discovery.product == EXPECTED_PRODUCT) { "mobile_platform_product_mismatch" }
        require(discovery.environment in setOf("development", "staging", "production")) {
            "mobile_platform_environment_invalid"
        }
        require(EXPECTED_PLATFORM_CONTRACT in discovery.supportedPlatformContracts) {
            "mobile_platform_contract_unsupported"
        }
        require(isSafeRelativeApiPath(discovery.capabilityEndpoint)) { "mobile_platform_capabilities_path_invalid" }
        require(discovery.authorizationIssuer.isNotBlank()) { "mobile_platform_authorization_issuer_required" }
        require(discovery.descriptorDigest.startsWith("sha256:")) { "mobile_platform_descriptor_digest_invalid" }
    }

    internal fun validateCapabilities(response: MobilePlatformCapabilityResponse) {
        require(response.maintenanceMode in setOf("none", "read_only", "unavailable")) {
            "mobile_platform_maintenance_mode_invalid"
        }
        require(response.platformContractMin.isNotBlank() && response.platformContractMax.isNotBlank()) {
            "mobile_platform_contract_range_required"
        }
        require(response.configDigest.startsWith("sha256:")) { "mobile_platform_config_digest_invalid" }
        require(response.validUntil.isNotBlank()) { "mobile_platform_valid_until_required" }
        response.capabilities.forEach { capability ->
            require(capability.capabilityId.isNotBlank()) { "mobile_platform_capability_id_required" }
            require(capability.version.isNotBlank()) { "mobile_platform_capability_version_required" }
            require(capability.state in setOf("available", "degraded", "read_only", "disabled")) {
                "mobile_platform_capability_state_invalid"
            }
            require(capability.validUntil.isNotBlank()) { "mobile_platform_capability_expiry_required" }
            require(!containsProviderResourceName(capability.capabilityId)) {
                "mobile_platform_provider_resource_detail_leaked"
            }
        }
    }

    internal fun isSafeRelativeApiPath(path: String): Boolean =
        path.startsWith('/')
            && !path.startsWith("//")
            && !path.contains("://")
            && !path.contains('\\')

    private fun containsProviderResourceName(value: String): Boolean {
        val normalized = value.lowercase()
        return listOf("r2_bucket", "d1_database", "durable_object", "hyperdrive_config", "queue_binding")
            .any(normalized::contains)
    }

    private fun getJson(url: String): JSONObject {
        val request = Request.Builder()
            .url(url)
            .get()
            .addHeader("Accept", "application/json")
            .addHeader("User-Agent", "ikimon-field-companion/platform-discovery")
            .build()
        client.newCall(request).execute().use { response ->
            val responseBody = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw IllegalStateException(responseBody.ifBlank { "mobile_platform_discovery_failed" })
            }
            return JSONObject(responseBody)
        }
    }
}
